src/protocol/basics/sum_of_product/malicious.rs

```diff
@@ -60,36 +60,36 @@ impl AsRef<str> for Step {
 pub async fn sum_of_products<F>(
     ctx: MaliciousContext<'_, F>,
     record_id: RecordId,
-    a: &[&MaliciousReplicated<F>],
-    b: &[&MaliciousReplicated<F>],
+    pairs: &[(&MaliciousReplicated<F>, &MaliciousReplicated<F>)],
```

Comment:
I thought this format might be more relatable since sum of products would be done on each pair. How does this affect the efficiency? (I am changing it in the code but trying to understand the efficiency consideration here)

Reply:
my assumption is that `a` and `b` are separate contiguous regions in memory (as previous implementation assumed). You somehow accumulated `a` values and `b` values before doing sum of products (maybe from two different sub-protocols). Once you hold onto `a` and `b` regions, you can just pass references to `a` and `b` to `sum_of_products`. Your implementation takes a slice of pairs, so clients that hold `a` and `b` will have to construct this new region (vector) of pairs by copying references to values inside `a` and `b`. That would require another allocation and therefore less efficient.

Reply:
You could achieve similar semantics without copying if you took an argument that was `impl IntoIterator<Item = (&M, &M)>`, but I don't know if that is any better.  If these really need to be held in a random access store in order to apply permutations, then there is no value in that change.

Reply:
Looking down below at how this is actually used - I don't think there's any difference because we *aren't* dealing with consecutive regions of memory. We are creating new vectors either way. So in practice I don't think there's a difference between creating two new vectors vs creating one vector of pairs.

What I like about this approach is that you have a compile-time guarantee that the vectors have the same length.

Reply:
I looked at the usages (I was only able to find usages in tests, so could be missing something) and it seems that `a` and `b` are coming as separate allocations most of the time?

for example

```rust
let res = world
            .semi_honest((av, bv), |ctx, (a, b)| async move {
                let mut pairs = Vec::with_capacity(MULTI_BIT_LEN);
                for i in 0..a.len() {
                    pairs.push((&a[i], &b[i]));
                }
                ctx.sum_of_products(RecordId::from(0), pairs.as_slice())
                    .await
                    .unwrap()
            })
```

with previous implementation can probably be written without another vector

```rust
let res = world
            .semi_honest((av, bv), |ctx, (a, b)| async move {
                ctx.sum_of_products(RecordId::from(0), &a, &b)
                    .await
                    .unwrap()
            })
```

I agree that ergonomic of using `&[(a, b)]` is better, but efficiency loss may be significant if we need to allocate another large vector just to combine them. If we are truly assembling the `a` and `b` vectors at the same time (within the same protocol), then I am in favor of this change